src/main/scala/devices/spi/SPIPhysical.scala

```diff
@@ -39,14 +46,46 @@ class SPIPhysical(c: SPIParamsBase) extends Module {
   val last = Wire(init = Bool(false))
   // Delayed versions
   val setup_d = Reg(next = setup)
-  val sample_d = ShiftRegInit(sample, c.sampleDelay, init = Bool(false))
-  val last_d = ShiftRegInit(last, c.sampleDelay, init = Bool(false))
 
   val scnt = Reg(init = UInt(0, c.countBits))
   val tcnt = Reg(io.ctrl.sck.div)
 
   val stop = (scnt === UInt(0))
   val beat = (tcnt === UInt(0))
+
+  //Making a delay counter for 'sample'
+  val del_cntr = RegInit(UInt(c.divisorBits.W), (io.ctrl.extradel.coarse + 1.U))
+  val sample_d = RegInit(Bool(false)) 
+  when (beat && sample){
+       del_cntr := io.ctrl.extradel.coarse
+    }
+  when  (del_cntr =/= 0.U){
+     (del_cntr := del_cntr - 1.U)
+  }
+  when (del_cntr === 1.U){
```

Comment:
should this be .elsewhen? if not, maybe you want to add a newline between these two to make it a little more clear...